File: packages/setpieces-py/src/setpieces/zones.py
"""Delivery-zone classification for corners and free kicks.

Deliveries are mirrored onto a single attacking side (near-post low,
far-post high) so left- and right-side set pieces combine into one
near/central/far-post x six-yard/edge-of-box 6-zone breakdown, plus a
"short" bucket for deliveries that don't reach the box at all.
"""

import collections
from dataclasses import dataclass
from typing import Dict, Iterable, Tuple, Union

from .events import SetPieceEvent

ZoneKey = Union[str, Tuple[str, str]]


@dataclass(frozen=True)
class ZoneParams:
    box_front_x: float = 83.0
    six_yard_front_x: float = 94.0
    near_cut: float = 41.0
    far_cut: float = 59.0


DEFAULT_ZONE_PARAMS = ZoneParams()


def classify_zone(end_x: float, end_y: float, start_y: float,
                   params: ZoneParams = DEFAULT_ZONE_PARAMS) -> ZoneKey:
    """Classify a single delivery's end location into a zone key.

    Returns "short" if the ball never reached the penalty box, otherwise
    a (near|central|far, six|edge) tuple.
    """
    if end_x < params.box_front_x:
        return "short"
    # mirror onto one side: deliveries from the right (start_y >= 50)
    # get their y flipped so "near post" always means the same thing
    my = 100 - end_y if start_y >= 50 else end_y
    col = "near" if my < params.near_cut else ("far" if my > params.far_cut else "central")
    row = "six" if end_x >= params.six_yard_front_x else "edge"
    return (col, row)


def zone_breakdown(events: Iterable[SetPieceEvent],
                    params: ZoneParams = DEFAULT_ZONE_PARAMS) -> Dict[ZoneKey, int]:
    """Raw counts per zone key across the given deliveries (events
    without an end location are skipped)."""
    counts: collections.Counter = collections.Counter()
    for e in events:
        if e.end_x is None or e.end_y is None:
            continue
        counts[classify_zone(e.end_x, e.end_y, e.y, params)] += 1
    return dict(counts)


def zone_percentages(events: Iterable[SetPieceEvent],
                      params: ZoneParams = DEFAULT_ZONE_PARAMS) -> Dict[ZoneKey, float]:
    """Each zone's share of all classified deliveries, as a percentage."""
    counts = zone_breakdown(events, params)
    total = sum(counts.values())
    if total == 0:
        return {}
    return {zone: n / total * 100.0 for zone, n in counts.items()}

File: packages/setpieces-r/setpieces/R/zones.R
# Delivery-zone classification for corners and free kicks.
#
# Deliveries are mirrored onto a single attacking side (near-post low,
# far-post high) so left- and right-side set pieces combine into one
# near/central/far-post x six-yard/edge-of-box 6-zone breakdown, plus a
# "short" bucket for deliveries that don't reach the box at all.

#' Default zone geometry parameters
#'
#' A named list with `box_front_x`, `six_yard_front_x`, `near_cut`, and
#' `far_cut`, all in Opta's 0-100 normalized pitch units.
#'
#' @export
default_zone_params <- list(
  box_front_x = 83.0,
  six_yard_front_x = 94.0,
  near_cut = 41.0,
  far_cut = 59.0
)

#' Classify a single delivery's end location into a zone key
#'
#' @param end_x,end_y Delivery end location (0-100 normalized).
#' @param start_y Y-coordinate the delivery was taken from (used to
#'   mirror right-side deliveries onto the left).
#' @param params Zone geometry parameters, see [default_zone_params].
#' @return `"short"` if the ball never reached the penalty box,
#'   otherwise a `"<near|central|far>_<six|edge>"` string.
#' @export
classify_zone <- function(end_x, end_y, start_y, params = default_zone_params) {
  if (is.na(end_x) || is.na(end_y)) return(NA_character_)
  if (end_x < params$box_front_x) return("short")

  my <- if (start_y >= 50) 100 - end_y else end_y
  col <- if (my < params$near_cut) "near" else if (my > params$far_cut) "far" else "central"
  row <- if (end_x >= params$six_yard_front_x) "six" else "edge"
  paste(col, row, sep = "_")
}

#' Raw counts per zone key across a set of deliveries
#'
#' @param events A list of set-piece event records (from
#'   [extract_set_pieces()]). Events without an end location are
#'   skipped.
#' @param params Zone geometry parameters, see [default_zone_params].
#' @return A named integer vector, zone key -> count.
#' @export
zone_breakdown <- function(events, params = default_zone_params) {
  zones <- vapply(events, function(e) {
    if (is.na(e$end_x) || is.na(e$end_y)) return(NA_character_)
    classify_zone(e$end_x, e$end_y, e$y, params)
  }, character(1))
  zones <- zones[!is.na(zones)]
  if (length(zones) == 0) return(stats::setNames(integer(0), character(0)))
  table(zones)
}

#' Each zone's share of all classified deliveries, as a percentage
#'
#' @inheritParams zone_breakdown
#' @return A named numeric vector, zone key -> percentage.
#' @export
zone_percentages <- function(events, params = default_zone_params) {
  counts <- zone_breakdown(events, params)
  total <- sum(counts)
  if (total == 0) return(stats::setNames(numeric(0), character(0)))
  counts / total * 100.0
}
